derive/src/parser.rs

```diff
@@ -0,0 +1,97 @@
+use crate::error::CompileError;
+use syn::{spanned::Spanned, FnArg, Ident, ImplItem, ImplItemMethod, ReturnType};
+
+/// A parameter. This doesn't used for modeling `&self` or `&mut self` parameters.
+#[derive(Clone)]
+pub struct Param {
+    /// A generated identifier used to name temporary variables
+    /// used for storing this parameter in generated code.
+    ///
+    /// This ident is used primary used for its' span.
+    pub ident: syn::Ident,
+}
+
+/// A function definition parsed from an impl block.
+pub struct FuncDef {
+    /// Assigned index of this function.
+    pub index: u32,
+    pub name: String,
+    /// The parameter of this function. This excludes the `&self` or `&mut self`.
+    pub params: Vec<Param>,
+    pub return_ty: Option<proc_macro2::Span>,
+}
+
+/// This is the core data structure which contains the list of all defined functions
+/// and the data required for the code generator (e.g. for implementing a trait).
+pub struct ImplBlockDef {
+    /// List of all defined external functions.
+    pub funcs: Vec<FuncDef>,
+    /// The generics required to implement a trait for this type.
+    pub generics: syn::Generics,
+    /// The type declaration to implement a trait, most typically
+    /// represented by a structure.
+    ///
+    /// E.g.: `Foo<'a>`, `()`
+    pub ty: Box<syn::Type>,
+}
+
+/// Parse an incoming stream of tokens into externalities definition.
```

Comment:
what is "externalities" in `wasmi` terms?

wouldn't even "externals" be a foreign term for WebAssembly?

Reply:
Yes, I'd agree that this term migrated from other projects here. What would you propose here?

Reply:
maybe "host bindings", "host functions" or simply "host" (if it later can be modified that way that it works not only with functions) to keep with WebAssembly terminology

also, the attribute name (now "derive_externals") might also reflect this

Reply:
yeah, but I'm afraid we are really too far down the road to do that. There is already `Externals` trait. But yeah would be up to change the doc.